src/librustc_typeck/check/wfcheck.rs

```diff
@@ -253,16 +253,55 @@ impl<'ccx, 'gcx> CheckTypeWellFormedVisitor<'ccx, 'gcx> {
         });
     }
 
+    fn check_auto_trait(&mut self,
+                                   trait_def_id: DefId,
+                                   span: Span)
+    {
+        let predicates = self.tcx().lookup_predicates(trait_def_id);
+
+        // If we must exclude the Self : Trait predicate contained by all
+        // traits.
+        let no_refl_predicates : Vec<_> =
```

Comment:
@jroesch well in that case change it to `&Self: Foo` :)

Anyway, to test the self-type... um I'm not sure what's the cleanest bit of code to do it, have to dig around a bit.